### PR TITLE
Allow SAS token to be refreshed

### PR DESCRIPTION
**Pull request by bsiegel**
Completes the engineering work for #160

We still need to add a sample azure function to show how to vend SAS tokens.

### CODE REVIEW

File: sdk/core/AzureCore/Source/Util/ConvertingInitializers.swift [hunk 1]
```diff
@@ -57,23 +57,64 @@ extension Date {
...
+            formatter.locale = Locale(identifier: "en_US_POSIX")
+            formatter.timeZone = TimeZone(abbreviation: "GMT")
+            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ"
+            return formatter
+        }()
     }
 
     public enum Format {
         case rfc1123
+        case iso8601
```
Comment (tjprescott) [inline sdk/core/AzureCore/Source/Util/ConvertingInitializers.swift:82]: Why are you not exposing `iso8601WithMilliseconds` here? Is there a way to specify the `WithMilliseconds` aspect with associated values?

File: sdk/core/AzureCore/Source/Util/ConvertingInitializers.swift [hunk 2]
```diff
@@ -57,23 +57,64 @@ extension Date {
...
         public var formatter: DateFormatter {
             switch self {
             case .rfc1123:
                 return Formatters.rfc1123
+            case .iso8601:
+                return Formatters.iso8601
+            }
+        }
+
+        fileprivate var alternateFormatters: [DateFormatter] {
```
Comment (tjprescott) [inline sdk/core/AzureCore/Source/Util/ConvertingInitializers.swift:93]: I think exposing `.iso8601(withMilliseconds: true)` would be conceptually simpler.

File: sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift [hunk 3]
```diff
@@ -29,23 +29,102 @@ import Foundation
...
+    internal var queueEndpoint: String?
+    internal var fileEndpoint: String?
+    internal var tableEndpoint: String?
+    internal var error: Error?
+    internal var sasToken: String? {
+        didSet { expirationDate = StorageSASCredential.expiration(forToken: sasToken) }
+    }
+
+
+    // Treat tokens that expire in <= 2 minutes as expired
```
Comment (tjprescott) [inline sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift:44]: What's the rationale for this? What is the typical expiration?

File: sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift [hunk 4]
```diff
@@ -29,23 +29,102 @@ import Foundation
...
+        self.blobSasUriProvider = nil
+        self.connectionStringProvider = connectionStringProvider
+        self.error = configureFrom(connectionString: connectionString)
+    }
+
+    /// Create a shared access signature credential from a container- or blob-level shared access signature.
+    /// - Parameters:
+    ///   - blobSasUri: The initial container- or blob-level shared access signature URI.
+    ///   - blobSasUriProvider: A closure that returns a container- or blob-level shared access signature URI.
+    public init(blobSasUri: String, blobSasUriProvider: (() -> String)? = nil) {
```
Comment (tjprescott) [inline sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift:96]: Is it possible to provide just a provider and not an initial value... where it would simply need to contact the provider initially? Same question for the connection string variety. Or is it that the initial connection string contains some important info that the connection string provider can't retrieve?

File: sdk/core/AzureCore/Source/Util/ConvertingInitializers.swift (hunk 2 above)
Comment (bsiegel) [reply sdk/core/AzureCore/Source/Util/ConvertingInitializers.swift:93]: I actually spent a lot of time on this, because it was annoying. In fact, one of the passes I took at it was exactly the suggestion you provided above. However I was trying to reconcile a few things that at first blush don't want to be reconciled:

- The SDK should be opinionated about how dates are formatted as strings
- The SDK must be flexible as to how to parse dates out of strings sent by the server
- The DateFormatter class (and the ISO8601DateFormatter class, which, incredibly, is *NOT* a subclass of DateFormatter) has no ability to be flexible when parsing strings to dates - the string MUST follow the precise format specified in the formatting string.

My attempt to resolve these constraints resulted in the solution you see here. It's asymmetric - the SDK can parse ISO8601 dates in two formats (note: this is still extremely insufficient to be a global solution - the [spec](https://en.wikipedia.org/wiki/ISO_8601) denotes a wide variety of formats that are valid ISO8601 date strings) but has an opinion about the only "correct" format in which a date should be rendered as an ISO8601 date string (and the opinion is that millisecond-accuracy is unnecessary, so we never emit fractional seconds).

Frustratingly, the ISO8601DateFormatter class exists that is aware of all of the different "options" you have available when formatting a date into an ISO8601 date string. But unfortunately it suffers the same limitation as DateFormatter, in that you must set the options to exactly match those present in the date string before parsing it into a Date, otherwise it will return nil even if the string is a valid ISO8601 date string.

I'm totally open to other suggestions here. If there were a way to accept *any* valid ISO8601 date string when parsing a string to a Date, I would much prefer that. And likewise if it could be encoded into the API elegantly, I would love to allow an option set of `ISO8601DateFormatter.Options` to be provided when formatting a Date to a string in ISO8601 format.

File: sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift (hunk 3 above)
Comment (bsiegel) [reply sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift:44]: So, I was under the impression that the server would provide a meaningful error response if the token was expired or otherwise invalid, but it seems that's not (always?) the case. In at least some cases, an expired token simply returns a 404 as if the entity did not exist at all. So the logic for determining whether the token is expired or not is (unfortunately) entirely based on the expiration date stamped into the token itself.

In order to have a greater chance of success and to prevent the customer from getting an unexpected 404 or other error response, we optimistically refresh the token just a little bit before it expires. I'm trying to recall where this value came from, but it was discussed last week and it was suggested that other SDKs used a value of two minutes in order to be as small as possible while mitigating (small amounts of) clock skew between the client and the server.

I would definitely want to get some feedback from other language SDK developers and/or the storage service developers on this.

File: sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift (hunk 4 above)
Comment (bsiegel) [reply sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift:96]: Yep, those are above. I couldn't collapse them into a single initializer each because then both parameters would be optional and a customer could choose to provide neither of them. So now you can either provide a credential provider closure, or a static credential and an optional credential provider to be used to refresh the credential.

Comment (bsiegel) [reply sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift:96]: I actually debated removing the version with the static credential entirely, since we know tokens WILL expire eventually, so a static credential will eventually require to be refreshed. But I figured there might be scenarios where a developer makes a token that expires far in the future because they simply don't want to ever have to deal with the complexity of refreshing it, so they treat it like a static key.

File: sdk/core/AzureCore/Source/Util/ConvertingInitializers.swift (hunk 2 above)
Comment (bsiegel) [reply sdk/core/AzureCore/Source/Util/ConvertingInitializers.swift]: I decided to remove the fallback logic for now. iOS's built-in `JSONDecoder.DateDecodingStrategy` exposes `iso8601` as a strategy which also only parses dates without milliseconds so I feel like there is at least some prior art. And right now I'm only using it for parsing the date in the SAS token which lacks the milliseconds part anyhow. If/when we need to support iso8601 dates with milliseconds or other less common formatting options we can revisit.

File: sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift (hunk 4 above)
Comment (tjprescott) [reply sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift:96]: Recommend reaching out to @adrianhall on this question.

Comment (bsiegel) [reply sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift:96]: Discussed, we will ditch these initializers and go with only the version that takes the callback.

**Comment by tjprescott** [comment]
@bsiegel is this PR ready to be merged?

**Comment by tjprescott** [comment]
@bsiegel what needs to be done to get this PR merged?

**Comment by bsiegel** [comment]
I think this is ready to go. I will add the sample function in a separate PR.

File: sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift [hunk 5]
```diff
@@ -29,23 +29,74 @@ import Foundation
...
-    internal let fileEndpoint: String?
-    internal let tableEndpoint: String?
-    internal let sasToken: String?
-    internal let error: Error?
+    internal let credentialProvider: () throws -> String
+    internal var blobEndpoint: String?
+    internal var queueEndpoint: String?
+    internal var fileEndpoint: String?
+    internal var tableEndpoint: String?
+    internal var error: Error?
```
Comment (tjprescott) [inline sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift:37]: Why are these changed from `var` to `let`?

Comment (bsiegel) [reply sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift:37]: Whenever we need to refresh the token, we mutate these properties so they have to be `var`s and the functions `clear()` and `refresh()` have to be marked as `mutating`.
